add backlight GPIO control

stripwax opened the pull request:
also tidy up the exit for the test_16x2 / test_20x4 examples

dbrgn commented on RPLCD/lcd.py:
```diff
@@ -149,8 +150,14 @@ def __init__(self, pin_rs=15, pin_rw=18, pin_e=16, pins_data=[21, 22, 23, 24],
                 Pin to start data read or write (E). Default: 16.
             pins_data:
-                List of data bus pins in 8 bit mode (DB0-DB7) or in 8 bit mode
+                List of data bus pins in 8 bit mode (DB0-DB7) or in 4 bit mode
```
Hehe, thanks for catching this :)

dbrgn commented on RPLCD/lcd.py:
```diff
@@ -130,6 +130,7 @@ class CharLCD(object):
     # Init, setup, teardown
 
     def __init__(self, pin_rs=15, pin_rw=18, pin_e=16, pins_data=[21, 22, 23, 24],
+                       pin_backlight=None, backlight_active_low=False,
```
I think we should change the name of `backlight_active_low` to `backlight_mode` to be in line with `numbering_mode`.

On line 111 we could then add a new enum:

```python
class BacklightMode(enum.Enum):
    active_high = 1
    active_low = 2
```

That makes it possible to create more obvious mode assignments:

```python
pin_backlight=None, backlight_mode=BacklightMode.active_high,
```

dbrgn commented on RPLCD/lcd.py:
```diff
@@ -182,13 +189,18 @@ def __init__(self, pin_rs=15, pin_rw=18, pin_e=16, pins_data=[21, 22, 23, 24],
                               d4=block2[0], d5=block2[1], d6=block2[2], d7=block2[3],
+                              backlight=pin_backlight,
                               mode=numbering_mode)
+        self.backlight_active_low=backlight_active_low
```
Please change this to `backlight_mode` according to the comment above, and use a space around the `=` sign in line with [PEP8](https://www.python.org/dev/peps/pep-0008/) :)

dbrgn commented on RPLCD/lcd.py:
```diff
@@ -182,13 +189,18 @@ def __init__(self, pin_rs=15, pin_rw=18, pin_e=16, pins_data=[21, 22, 23, 24],
         GPIO.setmode(self.numbering_mode)
         for pin in list(filter(None, self.pins))[:-1]:
             GPIO.setup(pin, GPIO.OUT)
+        if pin_backlight != None:
```
When comparing with `None`, the `is` operator should be used: `if pin_backlight is not None:`

dbrgn commented on RPLCD/lcd.py:
```diff
@@ -182,13 +189,18 @@ def __init__(self, pin_rs=15, pin_rw=18, pin_e=16, pins_data=[21, 22, 23, 24],
             GPIO.setup(pin, GPIO.OUT)
+        if pin_backlight != None:
+	    GPIO.setup(pin_backlight,GPIO.OUT)
+            GPIO.output(pin_backlight,GPIO.HIGH if backlight_active_low else GPIO.LOW)
```
With the new enum, this needs to be changed to

```python
GPIO.setup(pin_backlight, GPIO.OUT)
backlight_output_mode = GPIO.HIGH if backlight_mode is BacklightMode.active_low else GPIO.LOW
GPIO.output(pin_backlight, backlight_output_mode)
```

dbrgn commented on RPLCD/lcd.py:
```diff
@@ -450,6 +462,11 @@ def create_char(self, location, bitmap):
+        """Turn backlight GPIO on or off."""
+        if self.pins.backlight != None:
+            GPIO.output(self.pins.backlight, bool(value) ^ bool(self.backlight_active_low))
+
```
This would probably make more sense as a property, right after `display_enabled` (around line 331):

```python
    def _get_backlight_enabled(self):
        # We could probably read the current GPIO output state via sysfs, but
        # for now let's just store the state in the class
        if self.pins.backlight is None:
            raise ValueError('You did not configure a GPIO pin for backlight control!')
        return bool(self._backlight_enabled)

    def _set_backlight_enabled(self, value):
        if self.pins.backlight is None:
            raise ValueError('You did not configure a GPIO pin for backlight control!')
        if not isinistance(value, bool):
            raise ValueError('backlight_enabled must be set to ``True`` or ``False``.')
        self._backlight_enabled = value
        GPIO.output(self.pins.backlight, value ^ (self.backlight_mode is BacklightMode.active_low))

    backlight_enabled = property(_get_backlight_enabled, _set_backlight_enabled,
            doc='Whether or not to turn on the backlight.')
```

It's a more consistent API that allows you to simply assign `True` and `False` to `lcd.backlight_enabled`.

dbrgn replied:
For this to work, we need to initialize the `self._backlight_enabled` variable. I'd do that around line 179, right after `self.numbering_mode = ...`:

```python
self._backlight_enabled = True
```

dbrgn commented:
Thanks for this contribution! :smiley: I hope I did not overwhelm you with the change requests. But maintaining a clean and consistent API is the most important thing for easy usage in my opinion.

stripwax commented:
Only half a year later, I've returned.  I've incorporated your changes although in a slightly different way. Hopefully you'll agree what I changed made sense too!

dbrgn commented:
Thanks, that's cool. Last request: Could you run the PEP8 code standard tool over your changes and fix the things it nags about? (You can install it through `pip install pep8`. Or just look at the following output.)

```
$ pep8 --max-line-length 99 --ignore E127,E128,E265,E26
RPLCD/lcd.py:113:1: E302 expected 2 blank lines, found 1
RPLCD/lcd.py:119:1: E302 expected 2 blank lines, found 1
RPLCD/lcd.py:136:100: E501 line too long (107 > 99 characters)
RPLCD/lcd.py:200:28: E225 missing whitespace around operator
RPLCD/lcd.py:208:1: E101 indentation contains mixed spaces and tabs
RPLCD/lcd.py:208:1: W191 indentation contains tabs
RPLCD/lcd.py:208:30: E231 missing whitespace after ','
RPLCD/lcd.py:209:1: E101 indentation contains mixed spaces and tabs
RPLCD/lcd.py:210:35: E225 missing whitespace around operator
RPLCD/lcd.py:304:12: E713 test for membership should be 'not in'
RPLCD/lcd.py:320:12: E713 test for membership should be 'not in'
RPLCD/lcd.py:347:12: E713 test for membership should be 'not in'
```

(I should probably add that to the README or contributing information, sorry for forgetting abou that...)

Also, what's `dave.py`? :)

stripwax commented:
Hm, there shouldn't be any dave.py .. let me check. Also the pep8, which I hadn't heard of before.. thanks

On 08:10, 17 November 2015, Danilo Bargen wrote:
> Thanks, that's cool. Last request: Could you run the PEP8 code standard tool over your changes and fix the things it nags about? (You can install it through pip install pep8. Or just look at the following output.) $ pep8 --max-line-length 99 --ignore E127,E128,E265,E26 RPLCD/lcd.py:113:1: E302 expected 2 blank lines, found 1 RPLCD/lcd.py:119:1: E302 expected 2 blank lines, found 1 RPLCD/lcd.py:136:100: E501 line too long (107 > 99 characters) RPLCD/lcd.py:200:28: E225 missing whitespace around operator RPLCD/lcd.py:208:1: E101 indentation contains mixed spaces and tabs RPLCD/lcd.py:208:1: W191 indentation contains tabs RPLCD/lcd.py:208:30: E231 missing whitespace after ',' RPLCD/lcd.py:209:1: E101 indentation contains mixed spaces and tabs RPLCD/lcd.py:210:35: E225 missing whitespace around operator RPLCD/lcd.py:304:12: E713 test for membership should be 'not in' RPLCD/lcd.py:320:12: E713 test for membership should be 'not in' RPLCD/lcd.py:347:12: E713 test for membership should be 'not in' 
> 
> (I should probably add that to the README or contributing information, sorry for forgetting abou that...)
> 
> 
> Also, what's dave.py? :)
> 
> 
> —
> Reply to this email directly or > view it on GitHub> .> 
>

stripwax commented:
A lot of those pep8 nags are in your code.  I hope you don't seriously expect me to fix your code before you are willing to merge my fixes to your code .  Happy to fix any nags I introduced with my changes

stripwax commented:
pep8 --max-line-length 99 --ignore E127,E128,E265,E26

./setup.py:33:11: E123 closing bracket does not match indentation of opening bracket's line
./setup.py:34:5: E124 closing bracket does not match visual indentation
./test_16x2.py:67:100: E501 line too long (110 > 99 characters)
./test_16x2.py:103:100: E501 line too long (100 > 99 characters)
./test_16x2.py:111:100: E501 line too long (102 > 99 characters)
./test_20x4.py:67:100: E501 line too long (110 > 99 characters)
./test_20x4.py:99:100: E501 line too long (115 > 99 characters)
./test_20x4.py:107:100: E501 line too long (102 > 99 characters)
./test_20x4.py:125:14: E225 missing whitespace around operator
./test_20x4.py:128:1: W391 blank line at end of file
./RPLCD/__init__.py:5:1: W391 blank line at end of file
./RPLCD/enum.py:222:5: E123 closing bracket does not match indentation of opening bracket's line
./RPLCD/enum.py:261:5: E123 closing bracket does not match indentation of opening bracket's line
./RPLCD/lcd.py:307:12: E713 test for membership should be 'not in'
./RPLCD/lcd.py:323:12: E713 test for membership should be 'not in'
./RPLCD/lcd.py:350:12: E713 test for membership should be 'not in'

dbrgn commented:
Oh, sh**, I somehow mis-pasted the line of code... Of course I don't expect you to fix my things :) The test code and the setup.py don't really matter. What I wanted to write was this, which only targets the `lcd.py` file:

    pep8 --max-line-length 99 --ignore E127,E128,E265,E266 RPLCD/lcd.py

Thanks for the pull request!

dbrgn commented:
Thanks, merged in 7d3c5d41ef35009d663e4f34ddd525baf30ee1a2.

By the way, is an active low circuit more common than active high? Is this assuming you have a pull up resistor and a transistor that you pull down to enable the light?

dbrgn commented:
I'll close this, but I'm still interested in your answer :)

stripwax commented:
Thanks! :)
I don't know if active low or high is more common. Maybe more research required, perhaps depends on whether you're using off-the-shelf components or one of the various 'pi ready' kits. I used a very common off-the-shelf lcd module with a blue backlight. Backlight voltage is 4.5v-5v so I needed a switching transistor, but could do it either way (active low or high)

On 08:23, 18 November 2015, Danilo Bargen wrote:
> Thanks, merged in > 7d3c5d4> . 
> 
> By the way, is an active low circuit more common than active high? Is this assuming you have a pull up resistor and a transistor that you pull down to enable the light?
> 
> 
> —
> Reply to this email directly or > view it on GitHub> .> 
>

dbrgn commented:
OK, I guess we'll just leave it at the current default :)
